TestingLowerBounds/Testing/Binary.lean

```diff
@@ -250,38 +250,154 @@ lemma bayesBinaryRisk_eq (μ ν : Measure 𝒳) (π : Measure Bool) :
+    Nonempty (Subtype (@IsMarkovKernel 𝒳 𝒴 m𝒳 m𝒴)) := by
+  simp only [nonempty_subtype, Subtype.exists]
+  let y : 𝒴 := Classical.ofNonempty
+  refine ⟨kernel.const _ (Measure.dirac y), kernel.measurable (kernel.const 𝒳 _), ?_⟩
+  change IsMarkovKernel (kernel.const 𝒳 (Measure.dirac y))
+  exact kernel.isMarkovKernel_const
+
 lemma bayesBinaryRisk_self (μ : Measure 𝒳) (π : Measure Bool) :
-    bayesBinaryRisk μ μ π = min (π {true}) (π {false}) * μ Set.univ := by
+    bayesBinaryRisk μ μ π = min (π {false}) (π {true}) * μ Set.univ := by
   rw [bayesBinaryRisk_eq]
-  sorry
-
-lemma bayesBinaryRisk_le_min (μ ν : Measure 𝒳) (π : Measure Bool) :
-    bayesBinaryRisk μ ν π ≤ min (π {true} * μ Set.univ) (π {false} * ν Set.univ) := by
-  sorry
+  refine le_antisymm ?_ ?_
+  · let η : kernel 𝒳 Bool :=
+      if π {true} ≤ π {false} then (kernel.const 𝒳 (Measure.dirac false))
+        else (kernel.const 𝒳 (Measure.dirac true))
+    convert iInf_le_of_le η ?_
+    simp_rw [η]
+    convert iInf_le ?_ ?_ using 1
+    · split_ifs with h <;> simp [le_of_not_ge, h]
+    · split_ifs <;> exact kernel.isMarkovKernel_const
+  · calc
+      _ ≥ ⨅ κ, ⨅ (_ : IsMarkovKernel κ), min (π {false}) (π {true}) * (μ ∘ₘ κ) {false}
+          + min (π {false}) (π {true}) * (μ ∘ₘ κ) {true} := by
+        gcongr <;> simp
+      _ = ⨅ κ, ⨅ (_ : IsMarkovKernel κ), min (π {false}) (π {true}) * μ Set.univ := by
+        simp_rw [← mul_add, ← measure_union (show Disjoint {false} {true} from by simp)
+          (by trivial), (set_fintype_card_eq_univ_iff ({false} ∪ {true})).mp rfl,
+          Measure.comp_apply_univ]
+        rfl
+      _ = _ := by
+        rw [iInf_subtype']
+        convert iInf_const
+        exact nonempty_subtype_isMarkovKernel_of_nonempty
 
 lemma bayesBinaryRisk_dirac (a b : ℝ≥0∞) (x : 𝒳) (π : Measure Bool) :
     bayesBinaryRisk (a • Measure.dirac x) (b • Measure.dirac x) π
-      = min (π {true} * b) (π {false} * a) := by
-  rw [bayesBinaryRisk_eq]
-  have (κ : kernel 𝒳 Bool) [IsMarkovKernel κ] :
-      π {true} * ((b • Measure.dirac x) ∘ₘ κ) {false}
-        + π {false} * ((a • Measure.dirac x) ∘ₘ κ) {true}
-      = (π {true} * b) * κ x {false} + (π {false} * a) * κ x {true} := by
-    have (b : ℝ≥0∞) : (b • Measure.dirac x) ∘ₘ κ = b • κ x := by
-      ext s hs
-      simp only [Measure.bind_apply hs (kernel.measurable _), lintegral_smul_measure,
-        Measure.smul_apply, smul_eq_mul]
-      rw [lintegral_dirac']
-      exact kernel.measurable_coe _ hs
-    simp_rw [this]
-    simp only [Measure.smul_apply, smul_eq_mul, mul_assoc]
-  simp_rw [this]
-  sorry
+      = min (π {false} * a) (π {true} * b) := by
+  rw [bayesBinaryRisk_smul_smul, bayesBinaryRisk_self]
+  simp [lintegral_dirac]
+
+lemma bayesBinaryRisk_le_min (μ ν : Measure 𝒳) (π : Measure Bool) :
+    bayesBinaryRisk μ ν π ≤ min (π {false} * μ Set.univ) (π {true} * ν Set.univ) := by
+  let η : kernel 𝒳 Unit := kernel.const 𝒳 (Measure.dirac ())
```

Comment:
Can you use the `discard` kernel defined in Kernel/Monoidal.lean (in our project)? And add any missing API for it.